### ISSUE TITLE
Gnocchi equivalent for ceilometer /v2/meter/[meter]/statistics

### ISSUE THREAD

**Issue by briosheje** [gnocchixyz/gnocchi#1201]
I'm replacing Ceilometer with Gnocchi in an application but I'm having issues trying to get the same data that the ceilometer statistics api used to offer.

I'm starting from this example: 
```
/v2/meters/ip.floating.ext/statistics?q.op=eq&q.op=ge&q.op=lt&q.op=eq&q.value=[project_id]&q.value=2022-03-09T11:00:00&q.value=2022-03-09T14:00:00&q.value=no&q.field=project_id&q.field=timestamp&q.field=timestamp&q.field=metadata.in_use&period=3600
```

But I'm unable to find a way to get the same data in Gnocchi, neither from the `/metrics` api, neither with the list of metrics for resources.

Is there any way to get the same data as the above query from ceilometer?

I've tried digging in the documentation but even the aggregation API does not seem to allow to get such data.

**Comment by tobias-urdin** [comment]
Hello 👋 Can you give more details on how you want the data returned with some examples?
I assume you mean that you've read through the Dynamic Aggregates API documentation [1] otherwise that's a good start.

[1] https://gnocchi.osci.io/rest.html#dynamic-aggregates

**Comment by briosheje** [comment]
> Hello 👋 Can you give more details on how you want the data returned with some examples? I assume you mean that you've read through the Dynamic Aggregates API documentation [1] otherwise that's a good start.
> 
> [1] https://gnocchi.osci.io/rest.html#dynamic-aggregates

Hello!

Yes, I did read about the dynamic aggregates API, but I unluckily can't use that API because in the environment targeted for some reasons the Gnocchi API version is too old (and does not yet have the aggregates API, but does have to obsolete aggregation API).

I've found a way to use the aggregation API to group the metrics by what I need, but I didn't find any way to aggregate the data like the old API did.

Basically, the new API returns the measures grouped by the parameters you provide; the old API, instead, provides the grouped measures statistics (average, sum, count etc.) so to get this data it looks like I must aggregate the data manually in a second method.

What I was wondering is whether with the Aggregate api it's possible to perform such aggregations, but I didn't find any way to do that yet.

**Comment by tobias-urdin** [comment]
Hm, maybe you are looking for the reaggregation feature that is also available in the legacy API [1] that would allow you to do computation on the retrieved metrics?

[1] https://gnocchi.osci.io/rest.html?highlight=reaggregate#aggregation-across-metrics-deprecated

**Comment by tobias-urdin** [comment]
I kind of understand, hard to visualize without examples but what I think would be easiest for you since you cannot rely on any new functionality is to actually define a archive policy with the granularity and timespan you want and with sum aggregation computed etc, then simply retrieve that with or without a groupby statement.

Maybe something like that? I'm sorry that I cannot help you any further, I'm trying to get you further with solving it the best I can.

**Comment by briosheje** [comment]
> I kind of understand, hard to visualize without examples but what I think would be easiest for you since you cannot rely on any new functionality is to actually define a archive policy with the granularity and timespan you want and with sum aggregation computed etc, then simply retrieve that with or without a groupby statement.
> 
> Maybe something like that? I'm sorry that I cannot help you any further, I'm trying to get you further with solving it the best I can.

Indeed, in fact with some examples it might be clearer, let me try to give a better overview of what I'm trying to achieve.

Let's start from what the old API did: basically, starting from some filters and some grouping, a single record is returned which has the type marked as "statistics" in the old api reference: https://docs.openstack.org/ceilometer/pike/webapi/v2.html#Statistics

The statistics object is represented as follow (example):

```json
{
    "avg": 4.5,
    "count": 10,
    "duration": 300.0,
    "duration_end": "2013-01-04T16:47:00",
    "duration_start": "2013-01-04T16:42:00",
    "max": 9.0,
    "min": 1.0,
    "period": 7200,
    "period_end": "2013-01-04T18:00:00",
    "period_start": "2013-01-04T16:00:00",
    "sum": 45.0,
    "unit": "GiB"
}
```

Where, basically, starting from a period range (and date range), you get all the statistics of the aggregate measures (sum, count etc).

In Gnocchi, I was unable to find a way to produce the same output as above, the closest I could get is using either the aggregate or aggregation API to produce array of measures (filtered) that I later need to manually aggregate through external routines.

The core question, in a nutshell, is whether there is some way to produce the above output by delegating the calculation of the sum, average and other statistics to the API rather than doing that in a second moment.

Again, though, I have no problem doing the aggregation in a second moment and I would actually understand that a clean API should just return the raw results and is not meant to aggregate or bring statistics, so in either case it is fine; my main goal was to understand whether I did not find the answer in the documentation but there actually was one.

**Comment by tobias-urdin** [comment]
You won't be able to retrieve it the same way i.e the same output since you would need multiple queries to get different aggregated data (sum, min, max etc).

But I think what you are trying to do is simply a definition of an archive policy that matches your requirements. Then simply retrieving that with a start and stop timestamp.

**Comment by briosheje** [comment]
> You won't be able to retrieve it the same way i.e the same output since you would need multiple queries to get different aggregated data (sum, min, max etc).
> 
> But I think what you are trying to do is simply a definition of an archive policy that matches your requirements. Then simply retrieving that with a start and stop timestamp.

Ok, so as far as I can understand there are two possible ways:

1) perform X calls to the aggregate (or aggregation) api, one for each desired statistic (min, max, sum) and then aggregate the results manually. The count is (of course) the length of the returned measures array of either of the calls.
2) Define an archive policy that matches the requirement and perform X calls to the aggregate or aggregation api providing the previously defined archive policy.

Thanks
